Shasum check failed

Not sure if this is related to https://github.com/mixu/npm_lazy/issues/3 but we noticed that occasionally we come across a cached npm module that fails a checksum, here is a sample:

npm http GET https://npm_lazy.makerin.com/readable-stream/1.1.10
npm http 200 https://npm_lazy.makerin.com/readable-stream/1.1.10
npm http GET https://npm_lazy.makerin.com/readable-stream/-/readable-stream-1.1.10.tgz
npm http 200 https://npm_lazy.makerin.com/readable-stream/-/readable-stream-1.1.10.tgz
npm ERR! Error: shasum check failed for /home/ed/tmp/npm-27120-40gG7Kf9/1391030168947-0.10126865236088634/tmp.tgz
npm ERR! Expected: d4dc2e5319e9c90d1e71c69390ef62cd90827f65
npm ERR! Actual:   d40a009ef90e795283653b2fb8e601a26d83113b
npm ERR!     at /usr/local/lib/node_modules/npm/node_modules/sha/index.js:38:8
npm ERR!     at ReadStream.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/sha/index.js:85:7)
npm ERR!     at ReadStream.EventEmitter.emit (events.js:117:20)
npm ERR!     at _stream_readable.js:920:16
npm ERR!     at process._tickCallback (node.js:415:13)
npm ERR! If you need help, you may report this _entire_ log,
npm ERR! including the npm and node versions, at:
npm ERR!     http://github.com/isaacs/npm/issues

At that point our only option is to purge the entire cache, is there a process for purging individual modules?

Thanks

--- a/lib/resource.js
+++ b/lib/resource.js
@@ -111,8 +111,27 @@ Resource.prototype.getReadablePath = function(onDone) {
     if (self.type == 'tar' && self.exists()) {
       self.emit('fetch-cached');
       // is this a tarfile and is it in the index?
-      // yes: return readable stream
-      return onDone(null, self.exists());
+      // yes: check sha1 hash
+      var Package = require('./package.js');
+      return Package.getIndex(self.getPackageName(), function(err, data) {
+        if (err) {
+          return onDone(err, null);
+        }
+        expectedHash = verify.getSha(self.basename, data);
+        verify.check(self.exists(), function(err, actualHash) {
+          if (err) {
+            return onDone(err, null);
+          }
+          if (actualHash === expectedHash) {
+            return onDone(null, self.exists()); // return readable stream if file is good
+          }
+
+          // otherwise, cache is corrupted somehow, so bust cache and retry
+          console.log('Cached package is corrupt. Refetching ' + self.url);
+          Cache.junk(self.url);
+          self.getReadablePath(onDone);
+        });
+      });
     }
   }
 

